### ISSUE TITLE
Missing abyss-samtobreak-G3088269832 in ABySS 2.0 Makefile

### ISSUE THREAD

**Issue by hamzakhanvit** [sjackman/giab#2]
Hi,

From where can I get abyss-samtobreak-G3088269832, as mentioned in ABySS 2.0 [makefile](https://github.com/bcgsc/abyss-2.0-giab/blob/master/abyss2/Makefile) ?

**Comment by sjackman** [comment]
Hi, Hamza. If you checkout the `master` branch of https://github.com/bcgsc/abyss, install `ghc` with for example `sudo apt-get install ghc`, then build `abyss` from source, it will include the program `abyss-samtobreak`, which has an option `-G3088269832`.

**Comment by Christina-hshi** [comment]
Hi @sjackman ,
When running :  abyss-samtobreak -G3088269832  sam_file, I got an error,

> `Mapped NG50: abyss-samtobreak: samtobreak.hs:42:9-65: Irrefutable pattern failed for pattern Data.Maybe.Just (x, _)`

Do you know where is wrong? My sam file is generated after alignment with BWA.
BTW, it tooks me hours to install ghc. Maybe you can consider to offer us a binary version with compiling platform info attached for reference.

**Comment by sjackman** [comment]
Please report the output of `abyss-fac -G3088269832` on your assembly FASTA file. Is the assembly less than half you genome size? If so, that may produce the above error. Sorry for the ugly error message. If you omit the option `-G3088269832`, it'll calculate the NA50 rather than the NGA50, and avoid producing this error.

**Comment by sjackman** [comment]
> Maybe you can consider to offer us a binary version with compiling platform info attached for reference.

Good idea. I'll look into that.

**Comment by Christina-hshi** [comment]
@sjackman Thank you.  You are right that the error does occur when the assembly is less than half of the genome size. I changed the code a little bit. Now it works for me .

BTW, i want to share a little experience about how to use abyss-samtobreak to evaluate scaffolds. 
1. Firstly, for each scaffold, all consecutive regions with known bases are extracted as contigs (e.g. for scaffold NNATGTTCTNNGGAT, it contains two contigs, ATGTTCT and GGAT.), and each contig is named as scaffoldName_idForThisContig, where scaffoldName is the name of the scaffold where this contig is extracted and scaffoldName shouldn't contain any "_" inside. For example, I have two scaffolds, scaffoldA and scaffoldB, which contains two and three contigs inside. Then I can name them as scaffoldA_1, scaffoldA_2, scaffoldB_1, scaffoldB_2, and scaffoldB_3.
 2. Then, all extracted contigs are aligned to reference to produce a sam file, which will use as input of "abyss-samtobreak". 

@sjackman Am I right? Do I miss something?

**Comment by sjackman** [comment]
Yes, correct. To split up the FASTA file you can use `abyss-fatoagp -f scaftigs.fa`
or `seqtk cutN -n1 | tr : _`.
To align to the reference I recommend `bwa mem -Y -xintractg`.

**Comment by sjackman** [comment]
It sounds like you already have a fix for the NG50 bug. Here's another just in case:

```diff
    samtobreak: Fix NG50 error when assembly is small
    
    Fix an error when the assembly is smaller than half the genome size.
    Fix the error:
    Mapped NG50: abyss-samtobreak: ../../Misc/samtobreak.hs:42:9-65:
    Irrefutable pattern failed for pattern Just (x, _)

diff --git a/Misc/samtobreak.hs b/Misc/samtobreak.hs
index 3b577981..341e2d5a 100644
--- a/Misc/samtobreak.hs
+++ b/Misc/samtobreak.hs
@@ -10,6 +10,7 @@ import qualified Data.ByteString.Char8 as S
 import Data.Char (isDigit)
 import Data.Function (on)
 import Data.List (find, groupBy, intercalate, partition, sortBy, span)
+import Data.Maybe (fromMaybe)
 import System.Console.GetOpt
 import System.Environment (getArgs)
 import System.Exit (exitFailure, exitSuccess)
@@ -39,7 +40,7 @@ readS s = x where Just (x, _) = S.readInt s
 sumAtLeast :: Int -> [Int] -> Int
 sumAtLeast c xs = x
 	where
-	Just(x, _) = find ((>= c) . snd) $ zip ws $ scanl1 (+) ws
+	(x, _) = fromMaybe (0, 0) $ find ((>= c) . snd) $ zip ws $ scanl1 (+) ws
 	ws = sortBy (flip compare) xs
 
 -- Calculate NGx.
```

I'm currently working on other improvements to `abyss-samtobreak`, for example adding tab-separated-values (TSV) output format.

**Comment by sjackman** [comment]
> I changed the code a little bit. Now it works for me.

Have you coded in Haskell before? I'm duly impressed. `abyss-samtobreak` is the only real-world project that I've coded in Haskell. It's based on these two tools from Heng Li:
sam2break: https://github.com/lh3/fermi/blob/master/misc/sam2break.d
htsbox abreak: https://github.com/lh3/htsbox/blob/lite/abreak.c

I wrote `abyss-samtobreak` to calculate the scaffold NGA50.

**Comment by Christina-hshi** [comment]
@sjackman I haven't coded in Haskell before. Actually this is my first time to touch Haskell code. It takes me hours to read/understand the code. I really appreciate your efforts.  BTW, I am  wondering why don't you use other languages, e.g. Python or C++? As for me, Python and C++ are much easier to read and debug.

**Comment by sjackman** [comment]
It was an experiment and learning experience for me. The rest of ABySS is written in C++.
